### PR TITLE
Plotting statistics of binarized features

### PR DESCRIPTION
**Pull request by alexrogozin12**
Added tool for plotting statistics on float feature in model. It takes borders from model and plots mean value of target, mean value of prediction and number of objects over all bins. 

I hereby agree to the terms of the CLA available at: [link](https://yandex.ru/legal/cla/?lang=ru)

### CODE REVIEW

File: catboost/tutorials/model_analysis/binarized_feature_statistics_tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,280 @@
+{
```
Comment (annaveronika) [inline catboost/tutorials/model_analysis/binarized_feature_statistics_tutorial.ipynb]: The tutorial must be fully reproducible, and they must work fast (we tested on epsilon, because we wanted to be sure, it will work fast on large data).

So it would be better to use a well known open-sourced dataset for the tutorial.
You can choose user iris from sklearn for multiclassification:
from sklearn import datasets
iris = datasets.load_iris()
Titanic for one-hot encoding:
from catboost.datasets import titanic
And
boston = datasets.load_boston() for regression on numeric features.

File: catboost/python-package/catboost/core.py [hunk 2]
```diff
@@ -2060,6 +2061,45 @@ def set_params(self, **params):
             self._init_params[key] = value
         return self
 
+    def get_binarized_statistics(self, data, target, feature, prediction_type=None, cat_feature_values=[], plot=False):
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py]: `feature` -> `feature_index`
Also, let's add docstrings to this method like in other methods.

Comment (alexrogozin12) [reply catboost/python-package/catboost/core.py]: Here 'feature' maybe the name of feature, not only index

File: catboost/tutorials/model_analysis/binarized_feature_statistics_tutorial.ipynb (hunk 1 above)
Comment (alexrogozin12) [reply catboost/tutorials/model_analysis/binarized_feature_statistics_tutorial.ipynb]: Docs:

Gets statistics for the feature using the model, dataset and target.

For float features takes model's borders and computes
1) Mean target value across every bin;
2) Mean model prediction across every bin;
3) The number of objects in dataset which fall into each bin;
4) Predictions on varying feature. For every object, varies the feature value so
that it equals borders[0], borders[1], ... and counts model predictions. After that,
plots the average plot over all objects.

data -- numpy.array or pandas.DataFrame -- Data to compute statistics on -- Required parameter
target -- numpy.array or pandas.Series -- Target corresponding to data -- Required parameter
feature -- int or string -- Feature index or name in pd.DataFrame -- Required parameter
prediction_type -- str -- Prediction type used for counting mean_prediction: 'Class', 'Probability' or 'RawFormulaVal' -- None
cat_feature_values -- list or numpy.array or pandas -- Categorical feature values you need to get statistics on. Needed only for statistics of a categorical feature -- None
plot -- bool -- Plot statistics -- False

Type of return value
Python dict with binarized feature statistics. For float feature, includes
                'borders' -- borders for the specified feature in model
                'binarized_feature' -- numbers of bins where feature values fall
                'mean_target' -- mean value of target over each bin
                'mean_prediction' -- mean value of model prediction over each bin
                'objects_per_bin' -- number of objects per bin
                'predictions_on_varying_feature' -- averaged over dataset predictions for
                varying feature (see above)
            For one-hot feature, returns the same, but with 'cat_feature_values' instead of 'borders'

Comment (alexrogozin12) [reply catboost/tutorials/model_analysis/binarized_feature_statistics_tutorial.ipynb]: Usage example:


import catboost
import pandas as pd
from catboost.datasets import titanic
from plotly.offline import init_notebook_mode
init_notebook_mode(connected=True)

train, test = titanic()
target = train.Survived
train.drop(['PassengerId', 'Survived', 'Name', 'Parch', 'Ticket', 'Cabin', 'Embarked'], 
            axis=1, inplace=True)

model = catboost.CatBoostClassifier(
    iterations=200, 
    cat_features=['Pclass', 'Sex', 'SibSp'], 
    one_hot_max_size=300)
model.fit(train, target, silent=True)

feature = 'Fare'
res = model.get_binarized_statistics(train, target, feature, plot=True)

Comment (alexrogozin12) [reply catboost/tutorials/model_analysis/binarized_feature_statistics_tutorial.ipynb]: image

File: catboost/libs/quantized_pool_analysis/quantized_pool_analysis.h [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    struct TBinarizedFloatFeatureStatistics {
+        TVector<float> Borders;
+        TVector<ui8> BinarizedFeature;
+        TVector<float> MeanTarget;
+        TVector<float> MeanPrediction;
+        TVector<size_t> ObjectsPerBin;
+        TVector<double> PredictionsOnVaryingFeature;
+    };
+
+    struct TBinarizedOneHotFeatureStatistics{
```
Comment (nikitxskv) [inline catboost/libs/quantized_pool_analysis/quantized_pool_analysis.h]: Let's unite TBinarizedFloatFeatureStatistics and TBinarizedOneHotFeatureStatistics:
```
struct TFeatureStatistics {
    TVector<float> Borders; // Maybe empty in case of one hot features.
    TVector<int> BinarizedFeature;
    TVector<float> MeanTarget;
    TVector<float> MeanPrediction;
    TVector<size_t> ObjectsPerBin;
    TVector<double> PredictionsOnVaryingFeature;
};
```

File: catboost/libs/quantized_pool_analysis/quantized_pool_analysis.cpp [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        TRawDataProvider rawDataProvider(
+            std::move(dataProvider.MetaInfo),
+            std::move(dynamic_cast<TRawObjectsDataProvider*>(dataProvider.ObjectsData.Get())),
+            dataProvider.ObjectsGrouping,
+            std::move(dataProvider.RawTargetData));
+        size_t objectCount = rawDataProvider.GetObjectCount();
+        TIntrusivePtr<TRawDataProvider> rawDataProviderPtr(&rawDataProvider);
+        TRawBuilderData data = TRawBuilderDataHelper::Extract(std::move(*(rawDataProviderPtr.Get())));
+        THolder<THolderType> initialHolder;
+        if constexpr (FeatureType == EFeatureType::Float) {
```
Comment (nikitxskv) [inline catboost/libs/quantized_pool_analysis/quantized_pool_analysis.cpp]:
```
if constexpr (FeatureType == EFeatureType::Float) {
    initialHolder = std::move(data.ObjectsData.FloatFeatures[featureNum]);
} else {
    CB_ENSURE_INTERNAL(FeatureType == EFeatureType::Categorical, "Unsupported FeatureType");
    initialHolder = std::move(data.ObjectsData.CatFeatures[featureNum]);
}
```
And similarly below in both cases.

**Comment by nikitxskv** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
CLA already signed

**Comment by arcadia-devtools** [comment]
@nikitxskv, internal review request created: [822031](https://a.yandex-team.ru/review/822031)

**Comment by arcadia-devtools** [comment]
Check [822031](https://a.yandex-team.ru/review/822031) failed: 119 build(s) broken, 241 test(s) broken

**Comment by nikitxskv** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@nikitxskv, internal review request updated: [822031](https://a.yandex-team.ru/review/822031)

**Comment by arcadia-devtools** [comment]
Check [822031](https://a.yandex-team.ru/review/822031) failed: 24 test(s) broken

**Comment by nikitxskv** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@nikitxskv, internal review request updated: [822031](https://a.yandex-team.ru/review/822031)

**Comment by arcadia-devtools** [comment]
Check [822031](https://a.yandex-team.ru/review/822031) failed: 12 test(s) broken

**Comment by nikitxskv** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@nikitxskv, internal review request updated: [822031](https://a.yandex-team.ru/review/822031)

**Comment by arcadia-devtools** [comment]
Check [822031](https://a.yandex-team.ru/review/822031) failed: 25 build(s) broken, 6 test(s) broken

**Comment by nikitxskv** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@nikitxskv, internal review request updated: [822031](https://a.yandex-team.ru/review/822031)

**Comment by arcadia-devtools** [comment]
Check [822031](https://a.yandex-team.ru/review/822031) passed

File: catboost/python-package/catboost/core.py [hunk 5]
```diff
@@ -2074,6 +2075,106 @@ def set_params(self, **params):
...
+            if self.feature_names_ is None or feature not in self.feature_names_:
+                raise CatBoostError('No feature named "{}" in model'.format(feature))
+            if feature not in data.get_feature_names():
+                raise CatBoostError('No feature named "{}" in dataset'.format(feature))
+            feature = self.feature_names_.index(feature)
+
+        if prediction_type not in ['Class', 'Probability', 'RawFormulaVal']:
+            raise CatBoostError('Unknown prediction type "{}"'.format(prediction_type))
+
+        feature_type, feature_internal_index = self._object._get_feature_type_and_internal_index(feature)
```
Comment (nikitxskv) [inline catboost/python-package/catboost/core.py]: Here we can remove the duplication of code:
```
feature_type, feature_internal_index = self._object._get_feature_type_and_internal_index(feature)
feature_statistics = self._object._get_binarized_statistics(
    data,
    feature_internal_index,
    prediction_type,
    feature_type,
    thread_count
)

if feature_type == 'categorical':
    ...
```

**Comment by nikitxskv** [comment]
@arcadia-devtools Ship it!

**Comment by arcadia-devtools** [comment]
@nikitxskv, internal review request updated: [822031](https://a.yandex-team.ru/review/822031)

**Comment by arcadia-devtools** [comment]
Check [822031](https://a.yandex-team.ru/review/822031) passed

**Comment by annaveronika** [comment]
Thank you very much @alexrogozin12, this is an awesome contribution!
